src/couch_replicator/README.md

```diff
@@ -0,0 +1,295 @@
+   an `after` clause in receive and bound the maximum time this workerw will
+   take.
+
+   A doc processor worker will either succeed or fail but never block for too
+   long. Success and failure are returned as exit values. Those are handled in
+   the `worker_returned/3` doc processor clauses. The most common pattern is
+   that a worker is spawned to add a replication job, it does so and returns a
+   `{ok, ReplicationID}` value in `worker_returned`.
+
+   In case of a filtered replication with custom user code there are two case to
+   consider:
+
+     1. Filter fetching code has failed. In that case worker returns an error.
+        But because the error could be a transient network error, another
+        worker is started to try again. It could fail and return error again,
+        then another one is started and so on. However each consecutive worker
+        will do an exponential backoff, not unlike the scheduler code.
+        `error_backoff/1` is where the backoff period is calculated.
+        Consecutive errors are held in the `errcnt` field in the ETS table.
+
+     2. Fetchig filter code succeeds, replication ID is calculated and job is
+        added to the scheduler. However, because this is a filtered replication
+        source database could get an updated filter. Which means replication ID
+        should change again. So a worker is spawned again even if worker just
+        successfully returned successfully. The purpose is to check the filter
+        and see if it changed. So in other words doc processor will to do the
+        work of checking of filtered replications, get an updated filter and
+        will then refresh the replication job (remove the old one and add a new
+        one with a different ID). Filter checking interval is determined by the
+        `filter_backoff` function. An unusual thing about that function is that
+        it calculates the period based on the size of the ETS table. The
+        intuition is when there are few replications in a cluster, it's ok
+        to check the filter for changes often. When there are lots of
+        replications running, having each one checking their filter often is
+        not a good idea.
+
+ * `couch_replicator`: This is an unusual but useful pattern. This child is not
+   an actual process but a one-time call to the
+   `couch_replicator:ensure_rep_db_exists/0` function, executed by the
+   supervisor in the correct order (and monitored for crashes). This ensures
+   the local replicator db exists, then returns `ignore`. This pattern is
+   useful for doing setup-like things at the top level and in the correct order
+   regaring the rest of the children in the supervisor.
+
+ * `couch_replicator_db_changes`: This process specializes and configure
+   `couch_multidb_changes` so that it looks for `_replicator` suffixed shards
+   and makes sure to restart when cluster configuration changes. This restart
```

Comment:
I would add "or node membership" assuming that it does restart on an actual config change. Otherwise I'd just say node membership.